typing/typedecl.ml

```diff
@@ -829,102 +833,170 @@ let check_abbrev env sdecl (id, decl) =
-    in
-    if rec_ok then () else
-    let parents = TypeSet.add ty parents in
-    match get_desc ty with
-    | Tconstr(p, tyl, _) ->
-        let to_check = to_check p in
-        if to_check then List.iter (check_subtype parents trace ty) tyl;
-        begin match Ctype.try_expand_once_opt env ty with
-        | ty' ->
-            check parents (Expands_to (ty, ty') :: trace) ty'
-        | exception Ctype.Cannot_expand ->
-            if not to_check then List.iter (check_subtype parents trace ty) tyl
-        end
-    | _ ->
-        Btype.iter_type_expr (check_subtype parents trace ty) ty
-  and check_subtype parents trace outer_ty inner_ty =
-      check parents (Contains (outer_ty, inner_ty) :: trace) inner_ty
+        match trace with
+        | (Expands_to (ty1, _) :: _) as trace
+          when (match Btype.get_constr_desc ty1 with
+                Tconstr (p,_,_) -> Path.same p path | _ -> false) ->
+            List.rev trace, true
+        | trace -> List.rev trace, false
+      in
+      if rec_abbrev
+      then Recursive_abbrev (Path.name path, abs_env, reaching_path)
+      else Cycle_in_def (Path.name path, abs_env, reaching_path)
+    in raise (Error (loc, err))
+  in
+  let rec unguarded ~trace ty' =
+    if TypeSet.mem ty' !visited
+    then ()
+    else begin
+      match ty_opt with
+      | Some ty when eq_type ty ty' -> raise_error ~trace
+      | _ ->
+          match get_desc ty' with
+          | Tconstr (path, _, _)
+          | Texpand (_, path, _) when Path.same path ty_path ->
+              raise_error ~trace
+          | _ -> unguarded_no_self ~trace ty'
+    end
+  and unguarded_no_self ~trace ty' =
+    visited := TypeSet.add ty' !visited;
+    reachable
+      ~env:decl_env
+      ~rectypes_guarded
+      ~unguarded
+      ~trace
+      ty'
+  and rectypes_guarded ~trace ty' =
+    if !Clflags.recursive_types
+    then ()
+    else unguarded ~trace ty'
   in
+  if include_direct
+  then unguarded ~trace from_ty
+  else unguarded_no_self ~trace from_ty
+
+let is_reachable
+    ?trace
+    ~abs_env
+    ~decl_env
+    loc
+    ~from_ty
+    ~include_direct
+    ty_path
+    ty
+  =
   let snap = Btype.snapshot () in
-  try Ctype.wrap_trace_gadt_instances env (check TypeSet.empty []) ty0
+  try
+    Ctype.wrap_trace_gadt_instances decl_env (
+      is_reachable
+        ?trace
+        ~abs_env
+        ~decl_env
+        loc
+        ~from_ty
+        ~include_direct
+        ty_path
+    ) ty
   with Ctype.Escape _ ->
     (* Will be detected by check_regularity *)
     Btype.backtrack snap
 
-let check_well_founded_manifest ~abs_env env loc path decl =
+
+let check_well_founded_manifest ~abs_env loc path decl =
   if decl.type_manifest = None then () else
   let args = List.map (fun _ -> Ctype.newvar()) decl.type_params in
-  let visited = ref TypeMap.empty in
-  check_well_founded ~abs_env env loc path (Path.same path) visited
-    (Ctype.newconstr path args)
+  let ty = Ctype.newconstr path args in
```

Comment:
I think this currently is a no-op, since we can't expand `ty` in`abs_env`. However, do we even need this function?

It seems like the `check_well_founded_decl` function covers all the cases in the testsuite.

Reply:
Actually, turns out if you do the expansion in `new_env` (but still keep the abstract environment for the check), we prohibit
```ocaml
type 'a u = < x : 'a >
and 'a t = 'a t u
```
I'm thinking that we should not prohibit this type, since
```ocaml
type 'a u = < x : 'a >
type 'a t = 'a t u
```
is okay

So I would suggest removing `check_well_founded_manifest`